### PR TITLE
Add category and labels to automation/script save and rename dialog

### PR DESCRIPTION
**Pull request by jpbede**
## Proposed change
Add category and labels to the automation/script save and rename dialog. This way, a user can set the categories/labels at the time of creation and does not have to return to the list to do so.

Additionally this PR declutters the dialog, hides the optional parameters and make them addable via chips.
This is the first PR. There will be a second PR, when we've the possibility to catch the user when he wants to navigate away.

Worked closely with @matthiasdebaat on Discord on this, the redesign is based on his design.

WTH: https://community.home-assistant.io/t/wth-can-you-not-add-categories-labels-etc-when-creating-an-automation/802562

| image | image |
|:-:|:-:|
| image | image |

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Comment by bramkragten** [comment]
Nice! Should we also make the labels and categories visible on the device page? It feels weird to only have them visible in the list and now in the rename dialog.

**Comment by jpbede** [comment]
> Should we also make the labels and categories visible on the device page?

Yes, I like this.

**Comment by jpbede** [comment]
Changed the code of the save/rename dialog to be more generic, so we can use this as well for the scene editor.

File: src/panels/config/script/ha-script-editor.ts [hunk 1]
```diff
@@ -863,6 +901,9 @@ export class HaScriptEditor extends KeyboardShortcutMixin(LitElement) {
         ha-fab.dirty {
           bottom: 0;
         }
+        ha-fab.saving {
```
Comment (bramkragten) [inline src/panels/config/script/ha-script-editor.ts:904]: Doesnt the `disabled` property give it disabled styling? If not that should be fixed in `ha-fab`

File: src/panels/config/script/ha-script-editor.ts [hunk 2]
```diff
@@ -786,24 +802,46 @@ export class HaScriptEditor extends KeyboardShortcutMixin(LitElement) {
...
         this._config
       );
+
+      if (this._entityRegistryUpdate !== undefined) {
+        // wait for new script to appear in entity registry
+        if (!this.scriptId) {
+          await new Promise<void>((resolve, _reject) => {
+            setTimeout(resolve, 3000);
+          });
+        }
```
Comment (bramkragten) [inline src/panels/config/script/ha-script-editor.ts:819]: Not a fan of this...

File: src/panels/config/automation/ha-automation-editor.ts [hunk 3]
```diff
@@ -487,6 +512,17 @@ export class HaAutomationEditor extends KeyboardShortcutMixin(LitElement) {
         sub(this._config)
       );
     }
+
+    if (changedProps.has("entityRegistry")) {
+      const entry = this._getAutomationEntity(
+        this.entityRegistry,
+        this._entityId
+      );
+      this._entityRegistryUpdate = {
```
Comment (bramkragten) [inline src/panels/config/automation/ha-automation-editor.ts:521]: Wouldn't this overwrite the changes made by the user when the entity is created in the registry?

File: src/panels/config/script/ha-script-editor.ts (hunk 1 above)
Comment (jpbede) [reply src/panels/config/script/ha-script-editor.ts:904]: Will fix this in the `ha-fab`. The `disabled` does not give proper styling

File: src/panels/config/script/ha-script-editor.ts (hunk 2 above)
Comment (bramkragten) [reply src/panels/config/script/ha-script-editor.ts:819]: Also don't have a quick solution, a nicer option could be a promise that is resolved in the `willUpdate` handler when the entity registry is updated

Comment (jpbede) [reply src/panels/config/script/ha-script-editor.ts]: We need to delay the `updateEntityRegistryEntry` nevertheless, the `subscribeEntityRegistryUpdates` is debounced for 500ms. If we remove the `setTimeout` completely now, we would not get the updated entity entry, as it is debounced, which would result in a missing label/category on the list page until a complete reload occurs.

We could now directly subscribe the `entity_registry_updated` in the `ha-*-editor.ts`, bypassing the debounce, and unsubscribe after we received the confirmation that the entity has been created. What do you think?

Comment (jpbede) [reply src/panels/config/script/ha-script-editor.ts]: Nvm, we wouldn't get the second `entity_registry_updated` either, as the debounce would still be called, as other components are still subscribed.

File: src/panels/config/automation/ha-automation-editor.ts [hunk 4]
```diff
@@ -490,6 +503,17 @@ export class HaAutomationEditor extends PreventUnsavedMixin(
         sub(this._config)
       );
     }
+
+    if (changedProps.has("entityRegistry") && this.automationId) {
+      const entry = this.entityRegistry.find(
+        (ent) =>
+          ent.platform === "automation" && ent.unique_id === this.automationId
+      );
+      this._entityRegistryUpdate = {
```
Comment (bramkragten) [inline src/panels/config/automation/ha-automation-editor.ts:512]: Just never overwrite this, this one will only contain the changes of the user. If it is not set, we fallback to the actual entity registry for display.

File: src/panels/config/automation/ha-automation-editor.ts [hunk 5]
```diff
@@ -775,13 +798,15 @@ export class HaAutomationEditor extends PreventUnsavedMixin(
...
-        updateConfig: (config) => {
+        updateConfig: (config, entityRegistryUpdate) => {
           this._config = config;
+          this._entityRegistryUpdate = entityRegistryUpdate;
           this._dirty = true;
           this.requestUpdate();
           resolve(true);
         },
         onClose: () => resolve(false),
+        entityRegistryUpdate: this._entityRegistryUpdate,
```
Comment (bramkragten) [inline src/panels/config/automation/ha-automation-editor.ts:809]:
```suggestion
        entityRegistry: this._entityRegistry,
        entityRegistryUpdate: this._entityRegistryUpdate,
```

File: src/panels/config/automation/ha-automation-editor.ts [hunk 6]
```diff
@@ -817,21 +842,50 @@ export class HaAutomationEditor extends PreventUnsavedMixin(
...
       await saveAutomationConfig(this.hass, id, this._config!);
+
+      if (this._entityRegistryUpdate !== undefined) {
+        let entityId = this._entityId;
+
+        // wait for automation to appear in entity registry when creating a new automation
+        if (!entityId) {
+          await new Promise<void>((resolve) => {
+            setTimeout(resolve, 1000);
+          });
```
Comment (bramkragten) [inline src/panels/config/automation/ha-automation-editor.ts:858]: Maybe something like this?

```suggestion
	  const automation = await new Promise((resolve) => {
	    this._entityRegCreated = resolve;
	  });
```

```ts
protected willUpdate(changedProps) {
  if (this._entityRegCreated && changedProps.has("entityRegistry")) {
    const automation = this.entityRegistry.find(
        (entity: EntityRegistryEntry) => entity.unique_id === id
      );
    if (automation) {
    	this._entityRegCreated(automation);
    	this._entityRegCreated = undefined;
    }
  }
}
```

Comment (jpbede) [reply src/panels/config/automation/ha-automation-editor.ts:858]: Yes, this what I had, worked well but the problem is, that the second `entity_registry_updated` of the `updateEntityRegistryEntry` will be missed, as `subscribeEntityRegistryUpdates` is debounced (https://github.com/home-assistant/frontend/blob/e9b2a8341142543941b0f5c366a8ab354c7d79d4/src/data/entity_registry.ts#L252).

In most cases the `updateEntityRegistryEntry` will be call just a few milliseconds after the automation save.

Comment (jpbede) [reply src/panels/config/automation/ha-automation-editor.ts:858]: Okay, then I'm going to add the suggestion 👍

File: src/panels/config/automation/automation-rename-dialog/dialog-automation-rename.ts [hunk 7]
```diff
@@ -46,6 +56,17 @@ class DialogAutomationRename extends LitElement implements HassDialog {
         `ui.panel.config.${this._params.domain}.editor.default_name`
       );
     this._newDescription = params.config.description || "";
+    this._entryUpdates = params.entityRegistryUpdate || {
+      labels: params.entityRegistryEntry?.labels || [],
+      category: params.entityRegistryEntry?.categories[params.domain] || "",
+    };
+
+    this._visibleOptionals = [
+      this._newDescription! ? "description" : "",
```
Comment (bramkragten) [inline src/panels/config/automation/automation-rename-dialog/dialog-automation-rename.ts:65]:
```suggestion
      this._newDescription ? "description" : "",
```

File: src/panels/config/automation/automation-rename-dialog/dialog-automation-rename.ts [hunk 8]
```diff
@@ -46,6 +56,17 @@ class DialogAutomationRename extends LitElement implements HassDialog {
...
+      this._newIcon! ? "icon" : "",
```
Comment (bramkragten) [inline src/panels/config/automation/automation-rename-dialog/dialog-automation-rename.ts:66]:
```suggestion
      this._newIcon ? "icon" : "",
```

File: src/panels/config/automation/automation-rename-dialog/dialog-automation-rename.ts [hunk 9]
```diff
@@ -66,15 +100,23 @@ class DialogAutomationRename extends LitElement implements HassDialog {
...
         @closed=${this.closeDialog}
-        .heading=${createCloseHeading(
-          this.hass,
-          this.hass.localize(
-            this._params.config.alias
-              ? "ui.panel.config.automation.editor.rename"
-              : "ui.panel.config.automation.editor.save"
-          )
-        )}
+        .heading=${true}
```
Comment (bramkragten) [inline src/panels/config/automation/automation-rename-dialog/dialog-automation-rename.ts:103]: Make this a string for a11y
```suggestion
        .heading=${this.hass.localize(
              this._params.config.alias
                ? "ui.panel.config.automation.editor.rename"
                : "ui.panel.config.automation.editor.save"
            )}
```

File: src/panels/config/automation/automation-rename-dialog/dialog-automation-rename.ts [hunk 10]
```diff
@@ -162,19 +269,26 @@ class DialogAutomationRename extends LitElement implements HassDialog {
...
-        alias: this._newName,
-        description: this._newDescription,
-        icon: this._newIcon,
-      });
+      this._params.updateConfig(
+        {
+          ...this._params.config,
+          alias: this._newName,
+          description: this._newDescription,
+          icon: this._newIcon,
```
Comment (bramkragten) [inline src/panels/config/automation/automation-rename-dialog/dialog-automation-rename.ts:279]: Move these to `_entryUpdates` too

File: src/panels/config/automation/automation-rename-dialog/dialog-automation-rename.ts [hunk 11]
```diff
@@ -162,19 +269,26 @@ class DialogAutomationRename extends LitElement implements HassDialog {
...
-      this._params.updateConfig({
-        ...this._params.config,
-        alias: this._newName,
-        description: this._newDescription,
-      });
+      this._params.updateConfig(
+        {
+          ...this._params.config,
+          alias: this._newName,
+          description: this._newDescription,
```
Comment (bramkragten) [inline src/panels/config/automation/automation-rename-dialog/dialog-automation-rename.ts:288]: Move these to `_entryUpdates` too

File: src/panels/config/automation/ha-automation-editor.ts [hunk 12]
```diff
@@ -883,21 +922,49 @@ export class HaAutomationEditor extends PreventUnsavedMixin(
...
+              this._entityRegCreated = resolve;
+            }
+          );
+          entityId = automation.entity_id;
+        }
+
+        if (entityId) {
+          await updateEntityRegistryEntry(this.hass, entityId, {
+            categories: {
+              automation: this._entityRegistryUpdate.category || "",
```
Comment (bramkragten) [inline src/panels/config/automation/ha-automation-editor.ts:948]:
```suggestion
              automation: this._entityRegistryUpdate.category || null,
```

File: src/panels/config/automation/ha-config-automation.ts [hunk 13]
```diff
@@ -30,6 +33,10 @@ class HaConfigAutomation extends HassRouterPage {
 
   @property({ attribute: false }) public automations: AutomationEntity[] = [];
 
+  @state()
+  @consume({ context: fullEntitiesContext, subscribe: true })
+  _entityReg!: EntityRegistryEntry[];
```
Comment (bramkragten) [inline src/panels/config/automation/ha-config-automation.ts:38]: You can also move this to `automation-editor`

**Review by bramkragten — COMMENTED** [review]
Looks good! Can be merged after small nits

**Review by bramkragten — APPROVED** [review]

**Pull request merged**